File: Fibonacci/Fibonacci.cpp
#include<iostream>
#include<climits>
using namespace std;

//  Dynamic Programming


//  A Memoized Solution
/*
int Fibonacci(int n ,int memo){
	int result;
	if(memo[n] != 0){
		return memo[n];
	}
	if(n==1 || n==2){
		result = 1;
	}
	else{
		result = Fibonacci(n-1,memo)+Fibonacci(n-2,memo);	
	}
	memo[n] = result;
	return result;
}
*/


// Bottom-Top Approach
unsigned long long int FibonacciBottomTop(int n){
 	unsigned long long int Array[n + 2];
    Array[0] = 0;
    Array[1] = 1;
    for(int i = 2; i <= n; i++){
       Array[i] = Array[i - 1] + Array[i - 2];
    }
    return Array[n];
}

int main(){

	int n; 
	cout<<"Enter the Series Fibonacci Step: ";
	cin>>n;
	// int *memo = new int(n);
	cout<<"Value => "<< FibonacciBottomTop(n)<<endl<<endl;
	// for(int i=0;i<n;i++){
	// 	cout<<memo[i];
	// // }
	// cout<<ULLONG_MAX;
	return 0;

}

File: Fibonacci/Fibonacci.py
def recursiveFibonacci(n, memo):
    if memo[n] is not None:
        return memo[n]
    if n == 1 or n == 2:
        result = 1
    else:
        result = recursiveFibonacci(n-1, memo) + recursiveFibonacci(n-2, memo)
    memo[n] = result
    return result


def BottomTopFibonacci(n):
    if n == 1 or n == 2:
        return 1
    bottom_up = [None] * (n+1)
    bottom_up[1] = 1
    bottom_up[2] = 1
    for i in range(3, n+1):
        bottom_up[i] = bottom_up[i-1] + bottom_up[i-2]
    return bottom_up[n]


def NormalFibonacci(n):
    if n == 1 or n == 2:
        result = 1
    else:
        result = NormalFibonacci(n-1) + NormalFibonacci(n-2)
    return result


if __name__ == "__main__":
	n = int(input("Enter the Fibonacci Step: "))
	memo = [None] * (n + 1) ## [None,None,None<none]
	print(BottomTopFibonacci(n))


## 1 1 2 3